File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
import sys
input = sys.stdin.readline
N = int(input()) # 숫자의 개수
numbers = list(map(int, input().split()))
operations = list(map(int, input().split()))
min_num = int(1e9)
max_num = int(-1e9)

def dfs(depth, total, plus, minus, multiply, divide):
    global max_num, min_num
    # base case
    if depth == N:
        max_num = max(max_num, total)
        min_num = min(min_num, total)
        return
    if plus:
        dfs(depth+1, total + numbers[depth], plus-1, minus, multiply, divide)
    if minus:
        dfs(depth+1, total - numbers[depth], plus, minus-1, multiply, divide)
    if multiply:
        dfs(depth+1, total * numbers[depth], plus, minus, multiply-1, divide)
    if divide:
        dfs(depth+1, int(total / numbers[depth]), plus, minus, multiply, divide - 1)

dfs(1, numbers[0], operations[0], operations[1], operations[2], operations[3])
print(max_num)
print(min_num)

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.java
import java.util.*;
import java.io.*;
public class Main {
    public static int n;
    public static int[] arr;
    public static int[] op;
    public static int max = -1000000001;
    public static int min = 1000000001;
    
    public static void dfs(int depth, int cur, int plus, int minus, int mult, int div) {
        if (depth == n) {
            max = Math.max(max, cur);
            min = Math.min(min, cur);
        }
        if (plus < op[0]) {
            dfs(depth + 1, cur + arr[depth], plus + 1, minus, mult, div);
        }
        if (minus < op[1]) {
            dfs(depth + 1, cur - arr[depth], plus, minus + 1, mult, div);
        }
        if (mult < op[2]) {
            dfs(depth + 1, cur * arr[depth], plus, minus, mult + 1, div);
        }
        if (div < op[3]) {
            if (cur > 0) {
                dfs(depth + 1, cur / arr[depth], plus, minus, mult, div + 1);
            } else {
                dfs(depth + 1, (cur*(-1) / arr[depth]) * (-1), plus, minus, mult, div + 1);
            }
        }
    }
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());

        StringTokenizer st = new StringTokenizer(br.readLine());
        arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        st = new StringTokenizer(br.readLine());
        op = new int[4];
        for (int i = 0; i < 4; i++) {
            op[i] = Integer.parseInt(st.nextToken());
        }

        dfs(1, arr[0], 0, 0, 0, 0);
        System.out.println(max);
        System.out.println(min);
    }
}